staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go

```diff
@@ -235,41 +235,64 @@ func (c *AvailableConditionController) sync(key string) error {
 	}
 	// actually try to hit the discovery endpoint when it isn't local and when we're routing as a service.
 	if apiService.Spec.Service != nil && c.serviceResolver != nil {
-		discoveryURL, err := c.serviceResolver.ResolveEndpoint(apiService.Spec.Service.Namespace, apiService.Spec.Service.Name)
-		if err != nil {
-			return err
+		attempts := 5
```

Comment:
why 5?

Reply:
> why 5?

Chosen not quite at random.  We run HA with three masters and I wanted a good chance of trying each of them.

Reply:
Maybe it would be better to modify the service resolver to return the current list of endpoints.

Reply:
@lavalamp See above, dns, service ips, and load balancers are all allowed.  The list may not have more than one endpoint, but still needs to be tried more than once

Reply:
Are we getting value from duplicate requests to the same discoveryURL? Would it make sense to get the list of urls and dedup?

Reply:
> Are we getting value from duplicate requests to the same discoveryURL? Would it make sense to get the list of urls and dedup?

Same question from jordan above.  The answer is yes.  The same URL can actually be backed by multiple destinations (kube-proxy IP mapping and DNS name mapping come to mind) and even if you have one, multiple requests to a flaky endpoint also have value.  Falling out of this list is severely disruptive to controllers like quota and lifecycle.

Reply:
@deads2k I think that I disagree with you here, the `c.serviceResolver.ResolveEndpoint` method is returning the endpoints addresses, that are the actual Pod addresses

Reply:
> @deads2k I think that I disagree with you here, the `c.serviceResolver.ResolveEndpoint` method is returning the endpoints addresses, that are the actual Pod addresses

that depends on whether `--enable-aggregator-routing` is set to true or not, right?

Reply:
It does depend.  This code also has to choose whether it is better for kube-apiservers to stop routing to any aggregated apiserver endpoint based on an intermittent failure or whether it is better to suffer some intermittent failures.  The impact to the cluster of a single failure causing a full outage was quite bad.

Reply:
> that depends on whether --enable-aggregator-routing is set to true or not, right?

I see, the enable-aggregator-routing thing is something I don't fully control , I have to check in more detail

> This code also has to choose whether it is better for kube-apiservers to stop routing to any aggregated apiserver endpoint 

agree, however IIUIC, the published status is unique for the cluster, but each apiserver has an independent view of the network, so what happens in a 3 HA apisever configuration where apiserver A,B can reach the endpoints but C can not reach the endpoints  (it can connect to etcd, ...)

Reply:
it seems that the resolver resolve services to endpoints if `enable-aggregator-routing`is set, so it should be hitting the endpoints directly.
https://github.com/kubernetes/kubernetes/blob/b40d9e8dd5de13a11eb5c56c8fdb7794047764d9/cmd/kube-apiserver/app/server.go#L635

I'm going to test this better in a real cluster, I just I want to understand it better, since I spent last 2 weeks with bugs related to this availability_controller  and I think we can improve a bit things or at least make this errors easier to understand to the admins